File: frontend/src/api/media.ts
import apiClient from '../lib/api'
import axios from 'axios'
import { Media, PresignedUploadResponse, UploadRequest } from '../types'

/**
 * Get all media for an event
 * @param eventId Event UUID
 * @param limit Maximum number of results (default: 50)
 * @param offset Pagination offset (default: 0)
 * @returns Promise with Media array
 */
export const getEventMedia = async (
  eventId: string,
  limit: number = 50,
  offset: number = 0
): Promise<Media[]> => {
  const response = await apiClient.get(`/api/v1/media/events/${eventId}/media`, {
    params: {
      limit,
      offset,
    },
  })
  return response.data
}

/**
 * Get a single media item by ID
 * @param mediaId Media UUID
 * @returns Promise with Media object
 */
export const getMedia = async (mediaId: string): Promise<Media> => {
  const response = await apiClient.get(`/api/v1/media/${mediaId}`)
  return response.data
}

/**
 * Get presigned URL for uploading media
 * @param data Upload request data
 * @returns Promise with presigned URL and media ID
 */
export const getUploadUrl = async (data: UploadRequest): Promise<PresignedUploadResponse> => {
  const response = await apiClient.post('/api/v1/media/upload-url', data)
  return response.data
}

/**
 * Confirm media upload and trigger face detection
 * @param mediaId Media UUID
 * @returns Promise with confirmation status
 */
export const confirmUpload = async (mediaId: string): Promise<{ status: string; media_id: string }> => {
  const response = await apiClient.post(`/api/v1/media/${mediaId}/confirm`)
  return response.data
}

/**
 * Upload file directly to Azure Blob Storage using presigned URL
 * @param uploadUrl Presigned URL from getUploadUrl
 * @param file File to upload
 * @returns Promise that resolves when upload is complete
 */
export const uploadToBlob = async (uploadUrl: string, file: File): Promise<void> => {
  await axios.put(uploadUrl, file, {
    headers: {
      'Content-Type': file.type,
      'x-ms-blob-type': 'BlockBlob',
    },
  })
}

/**
 * Delete media
 * @param mediaId Media UUID
 * @returns Promise with deletion confirmation
 */
export const deleteMedia = async (mediaId: string): Promise<{ message: string }> => {
  const response = await apiClient.delete(`/api/v1/media/${mediaId}`)
  return response.data
}

File: backend/app/api/endpoints/media.py
"""Media upload and management endpoints"""

from fastapi import APIRouter, Depends, HTTPException, status, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List
from app.core.database import get_db
from app.models.database import Media
from app.schemas.pydantic import MediaUploadRequest, MediaResponse, PresignedUploadResponse
from app.api.deps import verify_access_code
from app.services.azure_blob import generate_presigned_upload_url
from app.workers.face_processor import process_media_faces

router = APIRouter()


@router.post("/upload-url", response_model=PresignedUploadResponse)
async def get_upload_url(
    upload_request: MediaUploadRequest,
    db: AsyncSession = Depends(get_db),
):
    """Get presigned URL for direct upload to Azure Blob Storage"""
    # TODO: Verify access code or user permissions
    
    # Generate presigned URL
    upload_url, blob_url = await generate_presigned_upload_url(
        filename=upload_request.filename,
        content_type=upload_request.content_type,
    )
    
    # Create media record in pending state
    media = Media(
        event_id=upload_request.event_id,
        uploader_id=upload_request.uploader_id,
        blob_url=blob_url,
        filename=upload_request.filename,
        content_type=upload_request.content_type,
        file_size=upload_request.file_size,
        media_type="photo",  # TODO: Detect from content_type
        face_detection_status="pending",
    )
    
    db.add(media)
    await db.commit()
    await db.refresh(media)
    
    return {
        "upload_url": upload_url,
        "media_id": media.media_id,
        "blob_url": blob_url,
    }


@router.post("/{media_id}/confirm")
async def confirm_upload(
    media_id: str,
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db),
):
    """Confirm upload and trigger face detection"""
    from sqlalchemy import select
    from app.models.database import Media as MediaModel
    
    # Get media from database
    result = await db.execute(
        select(MediaModel).where(MediaModel.media_id == media_id)
    )
    media = result.scalar_one_or_none()
    
    if not media:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Media not found"
        )
    
    # Update status to processing
    media.face_detection_status = "processing"
    await db.commit()
    
    # Trigger face detection in background
    background_tasks.add_task(process_media_faces, media_id)
    
    return {
        "status": "processing",
        "media_id": media_id,
        "message": "Face detection started"
    }


@router.get("/events/{event_id}/media", response_model=List[MediaResponse])
async def list_event_media(
    event_id: str,
    limit: int = 50,
    offset: int = 0,
    db: AsyncSession = Depends(get_db),
):
    """Get all media for an event with pagination"""
    from sqlalchemy import select, desc
    from app.models.database import Media as MediaModel
    
    # Query media for the event
    query = (
        select(MediaModel)
        .where(MediaModel.event_id == event_id)
        .order_by(desc(MediaModel.created_at))
        .limit(limit)
        .offset(offset)
    )
    
    result = await db.execute(query)
    media_list = result.scalars().all()
    
    return media_list

